backend/drive/drive_activity.go

```diff
@@ -0,0 +1,240 @@
+
+// Returns the type of a target and an associated title.
+func parseTarget(target *driveactivity.Target) *driveactivity.DriveItem {
+	return target.DriveItem
+}
+
+// Returns the first target for a list of targets.
+func getTargetInfo(targets []*driveactivity.Target) (targetInfo *driveactivity.DriveItem) {
+	for _, target := range targets {
+		if targetInfo = parseTarget(target); targetInfo != nil {
+			return
+		}
+	}
+	return nil
+}
+
+func (f *Fs) getParent(target *driveactivity.DriveItem) *parent {
+	targetTitle := target.Title
+	q := createItemQueryRequest(parseItemID(target.Name), MoveQuery, "")
+	response, err := f.activitySvc.Activity.Query(q).Do()
+	if err != nil {
+		fs.Errorf(targetTitle, "Unable to retrieve list of moves: %v", err)
+	}
+	fs.Debugf(targetTitle, "Retrieved Moves : %d\n", len(response.Activities))
+
+	for _, activity := range response.Activities {
+		for _, action := range activity.Actions {
+			if action.Detail.Move != nil {
+				for _, addedParent := range action.Detail.Move.AddedParents {
+					parentItem := addedParent.DriveItem
+					return &parent{ID: parseItemID(parentItem.Name), Name: parentItem.Title}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (f *Fs) parseActivityActions(driveTarget *driveactivity.DriveItem,
+	actions []*driveactivity.Action) (invalidatedParents map[string]string) {
+
+	var (
+		invalidatedParent *parent
+	)
+	invalidatedParents = make(map[string]string)
+	for _, action := range actions {
+		target := driveTarget
+		if action.Target != nil && parseTarget(action.Target) != nil {
+			target = parseTarget(action.Target)
+		}
+		targetTitle := target.Title
+		if action.Detail.Delete != nil {
+			fs.Debugf(targetTitle, "Deleted")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		if action.Detail.Rename != nil {
+			fs.Debugf(targetTitle, "Renamed")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		if action.Detail.Restore != nil {
+			fs.Debugf(targetTitle, "Restored")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		/* TODO: Verify if this is needed
+		if action.Detail.Edit != nil {
+			fs.Debugf(targetTitle, "Edited")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		*/
+		/* TODO: Verify if this is needed
+		if action.Detail.PermissionChange != nil {
+			fs.Debugf(targetTitle, "Permissions Changed")
+			invalidatedParent = f.getParent(target)
+			invalidatedParents = addParent(invalidatedParents, invalidatedParent)
+		}
+		*/
+		if action.Detail.Move != nil {
+			fs.Debugf(targetTitle, "Moved")
+			moveAction := action.Detail.Move
+			for _, addedParent := range moveAction.AddedParents {
+				if addedParent.DriveItem == nil {
+					fs.Errorf(targetTitle, "Invalid Parent: %v\n", addedParent)
```

Comment:
This message should have a bit more context when we get reports about it on the forum! Eg "Invalid parent on Moved activity item"

It also needs to return or something here as it will crash if `addedParent` is `nil`.